nvflare/app_common/executors/launcher_executor.py

```diff
@@ -78,149 +77,128 @@ def __init__(
         if task_name not in supported_tasks:
             self.log_error(fl_ctx, f"Task '{task_name}' is not in supported tasks: {supported_tasks}")
-            return make_reply(ReturnCode.BAD_TASK_DATA)
+            return False
 
         current_round = shareable.get_header(AppConstants.CURRENT_ROUND, None)
         total_rounds = shareable.get_header(AppConstants.NUM_ROUNDS, None)
         if task_name == self._train_task_name:
             if current_round is None:
                 self.log_error(fl_ctx, "missing current round")
-                return make_reply(ReturnCode.EXECUTION_EXCEPTION)
+                return False
 
             if total_rounds is None:
                 self.log_error(fl_ctx, "missing total number of rounds")
-                return make_reply(ReturnCode.EXECUTION_EXCEPTION)
-
-        # if not launched yet
-        if not self._launch_once or not self._launched:
-            self._init_pipe_handler()
-            self.prepare_config_for_launch(shareable, fl_ctx)
-            launch_success = self._launch(task_name, shareable, fl_ctx, abort_signal)
-            if not launch_success:
-                self.log_error(fl_ctx, f"launch task ({task_name}): failed")
-                return make_reply(ReturnCode.EXECUTION_EXCEPTION)
-            self._launched = True
-            self.log_info(fl_ctx, f"External process for task ({task_name}) is launched.")
-
-        if self.pipe_handler is None:
-            self.log_error(fl_ctx, "pipe_handler is None")
-            return make_reply(ReturnCode.EXECUTION_EXCEPTION)
-
-        # pipe handler starts checking for heartbeats only after the 3rd party code has been launched
-        self.pipe_handler.start()
-        self.pipe_handler.wait_for_other_end(timeout=self._launch_timeout)
-
-        result = self._exchange(task_name, shareable, fl_ctx, abort_signal)
-        self._result_fl_model = None
-        self._result_metrics = None
-
-        if not self._launch_once or self._job_end:
-            req = Message.new_request(topic=Topic.END, data="END")
-            has_been_read = self.pipe_handler.send_to_peer(req, timeout=self._peer_read_timeout)
-            if self._peer_read_timeout and not has_been_read:
-                self.log_warning(
-                    fl_ctx,
-                    f"3rd party does not get END msg in {self._peer_read_timeout} secs!",
-                )
-                return make_reply(ReturnCode.EXECUTION_EXCEPTION)
-            launch_finish = self._wait_launch_finish(task_name, shareable, fl_ctx, abort_signal)
-            if not launch_finish:
-                return make_reply(ReturnCode.EXECUTION_EXCEPTION)
-            self._clear_state()
-            self.log_info(fl_ctx, f"Launched external process for task ({task_name}) is finished.")
+                return False
+        return True
 
-        return result
+    def check_output_shareable(self, task_name: str, shareable: Shareable, fl_ctx: FLContext) -> bool:
```

Comment:
Where is this called? It should be called in the "execute" method, right? But it's not called there.

Reply:
it is called in base class TaskExchanger's execute method.